### PR TITLE
5935 fixes spacing pixdim inplace change

### LINKED ISSUE #5935: Why the data shape are different through transform "spacingd.
### Discussed in https://github.com/Project-MONAI/MONAI/discussions/5934

<div type='discussions-op-text'>

<sup>Originally posted by **xxsxxsxxs666** February  3, 2023</sup>

    val_transforms = Compose(
        [
            LoadImaged(keys=["image", "label"]),
            EnsureChannelFirstd(keys=["image", "label"]),
            ScaleIntensityRanged(
                keys=["image"], a_min=-57, a_max=400,
                b_min=0.0, b_max=1.0, clip=True,
            ),
            CropForegroundd(keys=["image", "label"], source_key="image"),  # default: value>0
            Orientationd(keys=["image", "label"], axcodes="RAI"),
            Spacingd(keys=["image", "label"], pixdim=(
                0.35, 0.35, 0.5), mode=("bilinear", "nearest")),  # Now only Sequential_str, How to add spline
        ]
    )

The shape and metadata are the same between image and label
image

this problem only happen when I use cachedataset. Nothing went wrong when I use normal dataset. </div>

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #5935

### Description
assigning `out_d[idx]` in the `__call__` shouldn't change `self.pixdim` in Spacing

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Review by KumoLiu — APPROVED** [review]

File: monai/transforms/spatial/array.py [hunk 1]
```diff
@@ -553,7 +553,7 @@ def __call__(
             input_affine = np.eye(sr + 1, dtype=np.float64)
         affine_ = to_affine_nd(sr, convert_data_type(input_affine, np.ndarray)[0])
 
-        out_d = self.pixdim[:sr]
+        out_d = self.pixdim[:sr].copy()
```
Comment (Nic-Ma) [inline monai/transforms/spatial/array.py:556]: Please confirm whether we need `deepcopy` here.

Thanks.

**Review by Nic-Ma — APPROVED** [review]

File: monai/transforms/spatial/array.py (hunk 1 above)
Comment (wyli) [reply monai/transforms/spatial/array.py:556]: yes, that should be fine given self.pixdim a numpy array.

**Comment by myron** [comment]
I struggled to find this bug for many hours in monai1.1
Spacing / Resampling transform changed the resolution randomly , in my case when using min_pixdim, max_pixdim, and Dataloader with several workers,  each worker can potentially change the self.pixdim, and next image gets wrong result...
Thanks for fixing it.

**Comment by wyli** [comment]
> I struggled to find this bug for many hours in monai1.1 Spacing / Resampling transform changed the resolution randomly , in my case when using min_pixdim, max_pixdim, and Dataloader with several workers, each worker can potentially change the self.pixdim, and next image gets wrong result... Thanks for fixing it.

Thanks for reporting. I hope there's a method to document recent unreleased bug fixes, it may save you some time. Currently we are not really using this `unreleased` section of the changelog https://github.com/Project-MONAI/MONAI/blob/dev/CHANGELOG.md#unreleased cc@Nic-Ma

@xxsxxsxxs666 really helped a lot by creating a small script for reproducing the issue https://github.com/Project-MONAI/MONAI/issues/5935.

**Pull request merged**